File: examples/fastapi-app/alembic/versions/001_create_domain_models.py
"""Create domain models for the FastAPI example app.

Revision ID: 001
Create Date: 2026-02-12

"""

from alembic import op
import sqlalchemy as sa
from sqlalchemy.dialects.postgresql import JSONB, UUID

revision = "001"
down_revision = None
branch_labels = None
depends_on = None


def upgrade() -> None:
    op.create_table(
        "orders",
        sa.Column("id", sa.Integer(), primary_key=True, autoincrement=True),
        sa.Column("customer_email", sa.String(255), nullable=False),
        sa.Column("items", JSONB, nullable=False, server_default="[]"),
        sa.Column("total", sa.Numeric(10, 2), nullable=True),
        sa.Column(
            "status",
            sa.String(50),
            nullable=False,
            server_default="pending",
        ),
        sa.Column("task_uuid", UUID(as_uuid=True), nullable=True),
        sa.Column(
            "created_at",
            sa.DateTime(timezone=True),
            server_default=sa.func.now(),
            nullable=False,
        ),
        sa.Column(
            "updated_at",
            sa.DateTime(timezone=True),
            server_default=sa.func.now(),
            onupdate=sa.func.now(),
            nullable=False,
        ),
    )
    op.create_index("ix_orders_task_uuid", "orders", ["task_uuid"])
    op.create_index("ix_orders_status", "orders", ["status"])

    op.create_table(
        "analytics_jobs",
        sa.Column("id", sa.Integer(), primary_key=True, autoincrement=True),
        sa.Column("source", sa.String(100), nullable=False),
        sa.Column("dataset_url", sa.String(500), nullable=True),
        sa.Column(
            "status",
            sa.String(50),
            nullable=False,
            server_default="pending",
        ),
        sa.Column("task_uuid", UUID(as_uuid=True), nullable=True),
        sa.Column(
            "created_at",
            sa.DateTime(timezone=True),
            server_default=sa.func.now(),
            nullable=False,
        ),
        sa.Column(
            "updated_at",
            sa.DateTime(timezone=True),
            server_default=sa.func.now(),
            onupdate=sa.func.now(),
            nullable=False,
        ),
    )
    op.create_index("ix_analytics_jobs_task_uuid", "analytics_jobs", ["task_uuid"])
    op.create_index("ix_analytics_jobs_status", "analytics_jobs", ["status"])

    op.create_table(
        "service_requests",
        sa.Column("id", sa.Integer(), primary_key=True, autoincrement=True),
        sa.Column("user_id", sa.String(100), nullable=False),
        sa.Column("request_type", sa.String(100), nullable=False),
        sa.Column(
            "status",
            sa.String(50),
            nullable=False,
            server_default="pending",
        ),
        sa.Column("result", JSONB, nullable=True),
        sa.Column("task_uuid", UUID(as_uuid=True), nullable=True),
        sa.Column(
            "created_at",
            sa.DateTime(timezone=True),
            server_default=sa.func.now(),
            nullable=False,
        ),
        sa.Column(
            "updated_at",
            sa.DateTime(timezone=True),
            server_default=sa.func.now(),
            onupdate=sa.func.now(),
            nullable=False,
        ),
    )
    op.create_index("ix_service_requests_task_uuid", "service_requests", ["task_uuid"])
    op.create_index("ix_service_requests_status", "service_requests", ["status"])

    op.create_table(
        "compliance_checks",
        sa.Column("id", sa.Integer(), primary_key=True, autoincrement=True),
        sa.Column("order_ref", sa.String(100), nullable=False),
        sa.Column("namespace", sa.String(100), nullable=False),
        sa.Column(
            "status",
            sa.String(50),
            nullable=False,
            server_default="pending",
        ),
        sa.Column("task_uuid", UUID(as_uuid=True), nullable=True),
        sa.Column(
            "created_at",
            sa.DateTime(timezone=True),
            server_default=sa.func.now(),
            nullable=False,
        ),
        sa.Column(
            "updated_at",
            sa.DateTime(timezone=True),
            server_default=sa.func.now(),
            onupdate=sa.func.now(),
            nullable=False,
        ),
    )
    op.create_index(
        "ix_compliance_checks_task_uuid", "compliance_checks", ["task_uuid"]
    )
    op.create_index("ix_compliance_checks_status", "compliance_checks", ["status"])


def downgrade() -> None:
    op.drop_table("compliance_checks")
    op.drop_table("service_requests")
    op.drop_table("analytics_jobs")
    op.drop_table("orders")

File: examples/rails-app/db/migrate/001_create_domain_models.rb
class CreateDomainModels < ActiveRecord::Migration[7.1]
  def change
    enable_extension 'pgcrypto' unless extension_enabled?('pgcrypto')

    # E-commerce orders
    create_table :orders, id: :uuid do |t|
      t.string   :customer_email, null: false
      t.jsonb    :items,          null: false, default: []
      t.decimal  :total,          precision: 12, scale: 2
      t.string   :status,         null: false, default: 'pending'
      t.uuid     :task_uuid

      t.timestamps
    end

    add_index :orders, :customer_email
    add_index :orders, :status
    add_index :orders, :task_uuid, unique: true, where: 'task_uuid IS NOT NULL'

    # Analytics / data pipeline jobs
    create_table :analytics_jobs, id: :uuid do |t|
      t.string   :source,      null: false
      t.string   :dataset_url
      t.string   :status,      null: false, default: 'pending'
      t.uuid     :task_uuid

      t.timestamps
    end

    add_index :analytics_jobs, :source
    add_index :analytics_jobs, :status
    add_index :analytics_jobs, :task_uuid, unique: true, where: 'task_uuid IS NOT NULL'

    # Microservice orchestration requests
    create_table :service_requests, id: :uuid do |t|
      t.string   :user_id
      t.string   :request_type, null: false
      t.string   :status,       null: false, default: 'pending'
      t.jsonb    :result,       default: {}
      t.uuid     :task_uuid

      t.timestamps
    end

    add_index :service_requests, :user_id
    add_index :service_requests, :request_type
    add_index :service_requests, :status
    add_index :service_requests, :task_uuid, unique: true, where: 'task_uuid IS NOT NULL'

    # Compliance / team-scaling checks
    create_table :compliance_checks, id: :uuid do |t|
      t.string   :order_ref,  null: false
      t.string   :namespace,  null: false
      t.string   :status,     null: false, default: 'pending'
      t.uuid     :task_uuid

      t.timestamps
    end

    add_index :compliance_checks, :order_ref
    add_index :compliance_checks, :namespace
    add_index :compliance_checks, :status
    add_index :compliance_checks, :task_uuid, unique: true, where: 'task_uuid IS NOT NULL'
  end
end
